src/PaymentGateways/PayPalCommerce/PayPalCommerce.php

```diff
@@ -14,36 +20,108 @@
  *
  * @since 2.9.0
  */
-class PayPalCommerce implements PaymentGateway
+class PayPalCommerce extends PaymentGateway
 {
+    /**
+     * @deprecated
+     *
+     * Use getId or id function to access payment gateway id.
+     *
+     */
     const GATEWAY_ID = 'paypal-commerce';
 
     /**
-     * @inheritDoc
+     * @unreleased
+     *
+     * @param int $formId
+     * @param array $args
+     *
+     * @return string
+     */
+    public function getLegacyFormFieldMarkup($formId, $args)
+    {
+        ob_start();
+        give(AdvancedCardFields::class)->addCreditCardForm($formId);
+
+        return ob_get_clean();
+    }
+
+    /**
+     * @unreleased
+     *
+     * @return string
+     */
+    public static function id()
+    {
+        return 'paypal-commerce';
+    }
+
+    /**
+     * @unreleased
+     *
+     * @return string
      */
     public function getId()
     {
-        return self::GATEWAY_ID;
+        return self::id();
     }
 
     /**
-     * @inheritDoc
+     * @unreleased
+     *
+     * @return string
      */
     public function getName()
     {
         return esc_html__('PayPal Donations', 'give');
     }
 
     /**
-     * @inheritDoc
+     * @unreleased
+     *
+     * @return string
      */
     public function getPaymentMethodLabel()
     {
         return esc_html__('Credit Card', 'give');
     }
 
     /**
-     * @inheritDoc
+     * @unreleased
+     *
+     * @param GatewayPaymentData $paymentData
+     *
+     * @return GatewayCommand
+     * @throws Exception
+     */
+    public function createPayment(GatewayPaymentData $paymentData)
+    {
+        $paypalOrderId = give_clean($_POST['payPalOrderId']);
+
+        if ( ! $paypalOrderId) {
+            throw new PaymentGatewayException(esc_html('PayPal order id is missing.', 'give'));
+        }
+
+        $paypalOrder = give(PayPalOrder::class)->getOrder($paypalOrderId);
```

Comment:
Returning a `PayPalOrder` based on request data is a good candidate for boundary using an `Action` class.

Here is an example (truncated) from the Stripe Gateway:

```php
class GetPaymentMethodFromRequest
{
    public function __invoke( GatewayPaymentData $paymentData )
    {
        if (empty($_POST['give_stripe_payment_method'])) {
            throw new PaymentMethodException('Payment Method Not Found');
        }

        return new PaymentMethod( give_clean( $_POST['give_stripe_payment_method'] ) );
    }
}
```

Reply:
fixed: https://github.com/impress-org/givewp/compare/701eb615daf9...bdb384de4d1a